OsmAnd-java/src/main/java/net/osmand/router/network/NetworkRouteGpxApproximator.java

```diff
@@ -0,0 +1,300 @@
+	}
+
+	private double getDistance(GpxRoutePoint start, GpxRoutePoint end) {
+		return MapUtils.getDistance(start.lat, start.lon, end.lat, end.lon);
+	}
+
+	private NetworkRouteSegment getMatchingGpxSegments(GpxRoutePoint p1, GpxRoutePoint p2) {
+		List<NetworkRouteSegment> segments = new ArrayList<>();
+		for (NetworkRouteSegment segStart : p1.getObjects()) {
+			for (NetworkRouteSegment segEnd : p2.getObjects()) {
+				if (segEnd.getId() == segStart.getId() && segStart.start != segEnd.start) {
+					segments.add(new NetworkRouteSegment(segStart, segStart.start, segEnd.start));
+				}
+			}
+		}
+		//fix https://www.openstreetmap.org/way/51203425
+		NetworkRouteSegment res = null;
+		if (!segments.isEmpty()) {
+			double minLength = Double.MAX_VALUE;
+			for (NetworkRouteSegment segment : segments) {
+				double length = segment.robj.distance(segment.start, segment.end);
+				if (length < minLength) {
+					minLength = length;
+					res = segment;
+				}
+			}
+		}
+		return res;
+	}
+
+	private List<NetworkRouteSegment> loadDataByGPX(GPXFile gpxFile) throws IOException {
+		List<GpxRoutePoint> gpxRoutePoints = new ArrayList<>();
+		List<NetworkRouteSegment> res = new ArrayList<>();
+		List<NetworkRoutePoint> passedRoutePoints = new ArrayList<>();
+		int totalDistance = 0;
+		int unmatchedDistance = 0;
+		for (GPXUtilities.Track t : gpxFile.tracks) {
+			for (GPXUtilities.TrkSegment ts : t.segments) {
+				for (int i = 0; i < ts.points.size() - 1; i++) {
+					GPXUtilities.WptPt ps = ts.points.get(i);
+					NetworkRoutePoint nearesetPoint = selector.getNetworkRouteContext()
+							.getClosestNetworkRoutePoint(MapUtils.get31TileNumberX(ps.lon), MapUtils.get31TileNumberY(ps.lat));
+					GpxRoutePoint gpxRoutePoint = new GpxRoutePoint(ps.lat, ps.lon);
+					if (MapUtils.squareRootDist31(MapUtils.get31TileNumberX(ps.lon), MapUtils.get31TileNumberY(ps.lat),
+							nearesetPoint.x31, nearesetPoint.y31) < GPX_MAX_DISTANCE_POINT_MATCH) {
+						gpxRoutePoint.routePoint = nearesetPoint;
+					}
+					if (!passedRoutePoints.contains(nearesetPoint)) {
+						passedRoutePoints.add(nearesetPoint);
+						gpxRoutePoints.add(gpxRoutePoint);
+					}
+				}
+			}
+		}
+		for (int idx = 0; idx < gpxRoutePoints.size() - 1; idx++) {
+			GpxRoutePoint start = gpxRoutePoints.get(idx);
+			GpxRoutePoint nextPoint = gpxRoutePoints.get(idx + 1);
+			totalDistance += getDistance(start, nextPoint);
+			// 1. simple segment matching
+			NetworkRouteSegment matchingGpxSegment = getMatchingGpxSegments(start, nextPoint);
+			if (matchingGpxSegment != null) {
+				res.add(matchingGpxSegment);
+				continue;
+			}
+			// 2. skip extra gpx points
+			matchingGpxSegment = getGpxSegmentWithoutExtraGpxPoints(gpxRoutePoints, idx, start);
+			if (matchingGpxSegment != null) {
+				res.add(matchingGpxSegment);
+				continue;
+			}
+			unmatchedDistance += getDistance(start, nextPoint);
```

Comment:
Here unmatched Straight line could be added immediately without any sorting (similar to existing method) private void addStraightLine(GpxRouteApproximation gctx, List<LatLon> lastStraightLine, GpxPoint strPnt, RouteRegion reg) {)